src/NuGet.Core/NuGet.DependencyResolver.Core/ResolverUtility.cs

```diff
@@ -0,0 +1,424 @@
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Common;
+using NuGet.Frameworks;
+using NuGet.LibraryModel;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+
+namespace NuGet.DependencyResolver
+{
+    public static class ResolverUtility
+    {
+        public static Task<GraphItem<RemoteResolveResult>> FindLibraryCachedAsync(
+            ConcurrentDictionary<LibraryRangeCacheKey, Task<GraphItem<RemoteResolveResult>>> cache,
+            LibraryRange libraryRange,
+            NuGetFramework framework,
+            GraphEdge<RemoteResolveResult> outerEdge,
+            RemoteWalkContext context,
+            CancellationToken cancellationToken)
+        {
+            var key = new LibraryRangeCacheKey(libraryRange, framework);
+
+            return cache.GetOrAdd(key, (cacheKey) =>
+                FindLibraryEntryAsync(cacheKey.LibraryRange, framework, outerEdge, context, cancellationToken));
+        }
+
+        public static async Task<GraphItem<RemoteResolveResult>> FindLibraryEntryAsync(
+            LibraryRange libraryRange,
+            NuGetFramework framework,
+            GraphEdge<RemoteResolveResult> outerEdge,
+            RemoteWalkContext context,
+            CancellationToken cancellationToken)
+        {
+            GraphItem<RemoteResolveResult> graphItem = null;
+            var currentCacheContext = context.CacheContext;
+
+            // Try up to two times to get the package. The second
+            // retry will refresh the cache if a package is listed 
+            // but fails to download. This can happen if the feed prunes
+            // the package.
+            for (var i = 0; i < 2 && graphItem == null; i++)
+            {
+                var match = await FindLibraryMatchAsync(
+                    libraryRange,
+                    framework,
+                    outerEdge,
+                    context.RemoteLibraryProviders,
+                    context.LocalLibraryProviders,
+                    context.ProjectLibraryProviders,
+                    currentCacheContext,
+                    context.Logger,
+                    cancellationToken);
+
+                if (match == null)
+                {
+                    return CreateUnresolvedMatch(libraryRange);
+                }
+
+                try
+                {
+                    graphItem = await CreateGraphItemAsync(match, framework, currentCacheContext, context.Logger, cancellationToken);
+                }
+                catch (InvalidCacheProtocolException) when (i == 0)
+                {
+                    // 1st failure, invalidate the cache and try again.
+                    // Clear the on disk and memory caches during the next request.
+                    currentCacheContext = currentCacheContext.Clone();
+                    currentCacheContext.MaxAge = DateTimeOffset.UtcNow;
+                    currentCacheContext.RefreshMemoryCache = true;
```

Comment:
Consider encapsulating the above 3 lines into a factory method. This would make `RefershMemoryCache` truly immutable.